### PR TITLE
Telegram bot

### PR DESCRIPTION
**Pull request by lassemoldrup**
## Purpose

Implements #14.

## Changes

Implements a Telegram bot as described in #14. Also partially implements #17 so that the bot can use the API.

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.

### CODE REVIEW

File: examples/some-verifier/src/db.rs [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    const PLATFORM: Platform;
+}
+
+/// A trait that is implemented for the Platform enum to give some utility functons.
+trait DbName {
+    /// The name of the corresponding table.
+    fn table_name(&self) -> &'static str;
+    /// The name of the corresponding column.
+    fn column_name(&self) -> &'static str;
+}
```
Comment (abizjak) [inline examples/some-verifier/src/db.rs:16]: I don't get why you don't just implement these as methods on the `Platform` type.

THis is an awful lot of complexity and needless use of macros.

File: examples/some-verifier/src/db.rs [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+                tracing::error!("connection error: {}", e);
+            }
+        });
+
+        Ok(Self { client })
+    }
+
+    pub async fn get_accounts<P: DbPlatform>(&self, id: i64) -> DbResult<Vec<Account>> {
+        self.client
+            .query_one(
```
Comment (abizjak) [inline examples/some-verifier/src/db.rs:98]: Since you have this you should do `LIMIT 1` in the query.

File: examples/some-verifier/src/db.rs [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+            fn column_name(&self) -> &'static str {
+                match self {$(
+                    Platform::$name => concat!($db_name, "_id"),
+                )*}
+            }
+        }
+
+        fn accounts_from_row(row: Row) -> Vec<Account> {
+            vec![ $( Account::from_row::<$name>(&row)),* ]
```
Comment (abizjak) [inline examples/some-verifier/src/db.rs:42]: It's a bit odd to return a singleton vector which I think this ends up being.

File: examples/some-verifier/src/db.rs (hunk 2 above)
Comment (abizjak) [reply examples/some-verifier/src/db.rs:98]: It's a bit odd you use `query_one` but then return a vector of accounts. Is this ever going to be more than one?

File: examples/some-verifier/src/db.rs [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+    pub async fn get_revocation_status(&self, account: &Account) -> DbResult<bool> {
+        self.client
+            .query_one(
+                &format!(
+                    "SELECT id, revoked FROM {} WHERE id = $1",
+                    account.platform.table_name()
+                ),
+                &[&account.id],
+            )
```
Comment (abizjak) [inline examples/some-verifier/src/db.rs:117]: Won't this panic if the id does not exist? What prevents such a call?

Shouldn't you use `query_opt` and then handle the failure.

File: examples/some-verifier/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+            "/verifications/telegram/:id",
+            get(handle_get_verifications::<Telegram>),
+        )
+        .route(
+            "/verifications/discord/:id",
+            get(handle_get_verifications::<Discord>),
+        )
+        .with_state(state);
+
+    let socket = SocketAddr::new(IpAddr::V4(Ipv4Addr::new(0, 0, 0, 0)), app.api_port);
```
Comment (abizjak) [inline examples/some-verifier/src/main.rs:146]: The entire address should be taken as a `listen-address`.

File: examples/some-verifier/src/main.rs [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+            // or a revocation status for whatever reason, probably because the user
+            // does not have that platform
+            let verifications = future::join_all(futures)
+                .await
+                .into_iter()
+                .filter_map(|res| res.ok())
+                .collect();
+            Json(verifications)
+        }
+        Err(_) => Json(vec![]),
```
Comment (abizjak) [inline examples/some-verifier/src/main.rs:189]: This does not seem reasonable. Errors should be mapped to errors, they should not be ignored like this  silently.

Database errors will be silently ignored like this.

File: examples/some-verifier/src/main.rs [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+                } else {
+                    format!("{} {}", user.first_name(), user.last_name())
+                }
+            } else {
+                user.username().clone()
+            };
+            Ok(name)
+        }
+        Platform::Discord => {
+            const API_ENDPOINT: &'static str = "https://discord.com/api/v10";
```
Comment (abizjak) [inline examples/some-verifier/src/main.rs:221]: Define a top-level constant.

File: examples/some-verifier/src/main.rs [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+        ))
+        .build()?;
+
+    let mut worker = rust_tdlib::client::Worker::builder()
+        .with_auth_state_handler(AuthStateHandlerProxy::new_with_encryption_key("".into()))
+        .build()?;
+    worker.start();
+    let tdlib_client = worker.bind_client(tdlib_client).await?;
+
+    let client = reqwest::Client::new();
```
Comment (abizjak) [inline examples/some-verifier/src/main.rs:125]: This client should have (configurable by flags) strict timeouts set (both to connection and request).

File: examples/some-verifier/.gitignore [hunk 9]
```diff
@@ -0,0 +1 @@
+tddb
```
[the remark this answers is not in the document]
Comment (lassemoldrup) [reply examples/some-verifier/.gitignore:1]: That is a folder used by TDLib, I'll add a comment.

File: examples/some-verifier-lib/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct Verification {
+    pub platform: Platform,
+    pub username: String,
+    pub revoked: bool,
+}
```
[the remark this answers is not in the document]
Comment (lassemoldrup) [reply examples/some-verifier-lib/src/lib.rs:25]: Will be fixed, when #17 is eventually merged.

File: examples/some-verifier/README.md [hunk 11]
```diff
@@ -0,0 +1,42 @@
+# SoMe Web3 ID Verifier dApp and API
+
+Checks social media credentials issued by the SoMe Web3 ID Issuer.\
```
[the remark this answers is not in the document]
Comment (lassemoldrup) [reply examples/some-verifier/README.md:3]: Will be fixed, when #17 is eventually merged.

File: examples/some-verifier/README.md [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+Also serves an API to the social media bots, where they can query they verification status of users.
+
+Currently the two supported platforms are Telegram and Discord
+
+## Dependencies
+
+We use [TDLib](https://core.telegram.org/tdlib/) to look up Telegram usernames, so it needs to be installed.\
```
[the remark this answers is not in the document]
Comment (lassemoldrup) [reply examples/some-verifier/README.md:10]: Will be fixed, when #17 is eventually merged.

File: examples/some-verifier-lib/src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,25 @@
+use std::fmt::{self, Display, Formatter};
+
```
[the remark this answers is not in the document]
Comment (lassemoldrup) [reply examples/some-verifier-lib/src/lib.rs:2]: Will be fixed, when #17 is eventually merged.

File: examples/some-verifier/README.md [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+with
+```
+git clone --depth 1 --branch v1.8.0 https://github.com/tdlib/td.git
+```
+
+## API
+
+### GET ``/verifications/{platform}/{userId}``
+
+Repsonds with a list of `Verification`s in JSON.
```
[the remark this answers is not in the document]
Comment (lassemoldrup) [reply examples/some-verifier/README.md:26]: Will be fixed, when #17 is eventually merged.

File: examples/some-verifier/src/db.rs (hunk 1 above)
Comment (lassemoldrup) [reply examples/some-verifier/src/db.rs:16]: Will be fixed, when #17 is eventually merged.

File: examples/some-verifier/src/db.rs (hunk 3 above)
Comment (lassemoldrup) [reply examples/some-verifier/src/db.rs:42]: It returns a vector with one `Account` per platform (so Telegram and Discord right now)

File: examples/some-verifier/src/db.rs (hunk 4 above)
Comment (lassemoldrup) [reply examples/some-verifier/src/db.rs:117]: It just returns an error, but `query_opt` is probably more appropriate, yes.

Will be fixed, when #17 is eventually merged.

File: examples/some-verifier/src/main.rs (hunk 5 above)
Comment (lassemoldrup) [reply examples/some-verifier/src/main.rs:146]: Will be fixed, when #17 is eventually merged.

File: examples/some-verifier/src/main.rs (hunk 6 above)
Comment (lassemoldrup) [reply examples/some-verifier/src/main.rs:189]: Will be fixed, when #17 is eventually merged.

File: examples/some-verifier/src/main.rs [hunk 15]
```diff
@@ -0,0 +1,233 @@
...
+    }
+}
+
+/// Looks up the username of the given account.
+async fn get_username(state: &AppState, account: &Account) -> anyhow::Result<String> {
+    #[derive(Deserialize)]
+    struct DiscordUser {
+        username: String,
+        discriminator: String,
+    }
```
[the remark this answers is not in the document]
Comment (lassemoldrup) [reply examples/some-verifier/src/main.rs:199]: Will be fixed, when #17 is eventually merged.

File: examples/some-verifier/src/main.rs (hunk 7 above)
Comment (lassemoldrup) [reply examples/some-verifier/src/main.rs:221]: Will be fixed, when #17 is eventually merged.

File: examples/some-verifier/src/main.rs (hunk 8 above)
Comment (lassemoldrup) [reply examples/some-verifier/src/main.rs:125]: Will be fixed, when #17 is eventually merged.

**Comment by soerenbf** [comment]
Comments addressed in https://github.com/Concordium/concordium-web3id/pull/68

**Pull request closed without merging**